generator/camera_info_collector.hpp

```diff
@@ -0,0 +1,157 @@
+
+#include "routing/base/followed_polyline.hpp"
+#include "routing/routing_helpers.hpp"
+
+#include "indexer/data_source.hpp"
+#include "indexer/feature_data.hpp"
+#include "indexer/feature_decl.hpp"
+#include "indexer/scales.hpp"
+
+#include "coding/file_writer.hpp"
+#include "coding/write_to_sink.hpp"
+
+#include "geometry/point2d.hpp"
+#include "geometry/segment2d.hpp"
+
+#include "base/assert.hpp"
+#include "base/checked_cast.hpp"
+#include "base/logging.hpp"
+#include "base/string_utils.hpp"
+
+#include <cstdint>
+#include <limits>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "boost/optional.hpp"
+
+namespace generator
+{
+class CamerasInfoCollector
+{
+public:
+  // Don't touch the order of enum!
+  enum class CameraDirection
+  {
+    Unknown,
+    Forward,
+    Backward,
+    Both
+  };
+
+  class Header
+  {
+  public:
+    void SetVersion(uint32_t v) { m_version = v; }
+    void SetAmount(uint32_t n) { m_amount = n; }
+    uint32_t GetAmount() const { return m_amount; }
+
+    template <typename T>
+    void Serialize(T & sink)
+    {
+      WriteToSink(sink, m_version);
+      WriteToSink(sink, m_amount);
+    }
+
+    template <typename T>
+    void Deserialize(T & sink)
+    {
+      ReadPrimitiveFromSource(sink, m_version);
+      ReadPrimitiveFromSource(sink, m_amount);
+    }
+
+    bool IsValid() const { return m_version == CamerasInfoCollector::kLatestVersion; }
+
+  private:
+    uint32_t m_version = 0;
+    uint32_t m_amount = 0;
+  };
+
+  CamerasInfoCollector(std::string const & dataFilePath, std::string const & camerasInfoPath,
+                       std::string const & osmIdsToFeatureIdsPath);
+
+  static uint32_t constexpr kLatestVersion = 1;
+  static uint32_t constexpr kMaxCameraSpeedKmpH = std::numeric_limits<uint8_t>::max();
+
+  void Serialize(FileWriter & writer) const;
+
+  bool IsValid() const { return m_valid; }
+
+private:
+
+  struct Camera
+  {
+    // feature_id and segment_id, where placed camera.
+    // m_coef - coef from 0 to 1 where it placed at segment from the begin.
+    struct SegmentCoord
+    {
+      SegmentCoord() = default;
+      SegmentCoord(uint32_t fId, uint32_t sId, double k) : m_featureId(fId), m_segmentId(sId), m_coef(k) {}
+
+      uint32_t m_featureId = 0;
+      uint32_t m_segmentId = 0;
+      double m_coef = 0.0;
+    };
+
+    static double constexpr kEqualityEps = 1e-5;
+
+    Camera(m2::PointD const & center, uint8_t const maxSpeed, std::vector<SegmentCoord> && ways)
+      : m_center(center), m_maxSpeedKmPH(maxSpeed), m_ways(std::move(ways))
+    {
+    }
+
+    void ParseDirection()
+    {
+      // After some research we understood, that camera's direction is a random option.
+      // This fact was checked with random-choosed cameras and google maps 3d view.
+      // The real direction and the indicated in the .osm differed.
+      // So let's this function, space in mwm will be. And implementation won't, until
```

Comment:
let's what?
space in mwm will be what?

Reply:
=> `will be exists`

Reply:
will exist?
Ok, but what about "let's"?

Reply:
check another comment about it